### PR TITLE
Redirect gateway-like urls to ipfs://

### PR DESCRIPTION
**Pull request by cypt4**
Resolves https://github.com/brave/brave-browser/issues/21454 Urls in format of https://bafy.ipfs.gateway.io or https://gateway.io/ipfs/bafy are now redirected to ipfs:// scheme if x-ipfs-path header is received
Sec review :https://github.com/brave/security/issues/1063

Resolves 

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
"Redirect IPFS resources to the configured IPFS gateway" setting should open ipfs:// urls and, so, should work with IPFS resolve method "ASK"

### CODE REVIEW

File: components/ipfs/ipfs_utils.cc [hunk 1]
```diff
@@ -425,4 +425,41 @@ std::string GetRegistryDomainFromIPNS(const GURL& url) {
...
 
+absl::optional<GURL> TranslateToCurrentGatewayUrl(const GURL& url) {
+  if (!url.is_valid() || !url.SchemeIsHTTPOrHTTPS()) {
+    return absl::nullopt;
+  }
+
+  std::vector<std::string> host_parts = base::SplitStringUsingSubstr(
+      url.host(), ".", base::KEEP_WHITESPACE, base::SPLIT_WANT_ALL);
+
+  if (host_parts.size() > 2 && IsValidCID(host_parts.at(0)) &&
```
Comment (iefremov) [inline components/ipfs/ipfs_utils.cc:436]: having an example of what original url transforms to which final url would be handy

**Comment by cypt4** [comment]
> I'd suggest to ask someone from the sec team to look at the parsing code

Opened sec review https://github.com/brave/security/issues/1063

File: components/ipfs/ipfs_utils.cc [hunk 2]
```diff
@@ -425,4 +425,45 @@ std::string GetRegistryDomainFromIPNS(const GURL& url) {
...
+      host_parts.at(1) == "ipfs") {
+    GURL final_url = GURL("ipfs://" + host_parts.at(0) + url.path());
+    GURL::Replacements replacements;
+    replacements.SetQueryStr(url.query_piece());
```
Comment (yrliou) [inline components/ipfs/ipfs_utils.cc:444]: Shouldn't we do `replacements.SetRefStr(url.ref_piece());` too? Need to add test for this too if so.
Same question to the if block below.

File: browser/ipfs/test/ipfs_service_browsertest.cc [hunk 3]
```diff
@@ -1053,57 +1067,175 @@ IN_PROC_BROWSER_TEST_F(IpfsServiceBrowserTest, CannotLoadIPFSImageFromHTTP) {
...
 }
 
 IN_PROC_BROWSER_TEST_F(IpfsServiceBrowserTest,
-                       TopLevelAutoRedirectsOnWithQuery) {
+                       TopLevelAutoRedirectsOn_DoNotTranslateSimpleUrls) {
   ResetTestServer(
-      base::BindRepeating(&IpfsServiceBrowserTest::HandleEmbeddedSrvrRequest,
+      base::BindRepeating(&IpfsServiceBrowserTest::HandlePublicGatewayRequest,
                           base::Unretained(this)));
   browser()->profile()->GetPrefs()->SetBoolean(kIPFSAutoRedirectGateway, true);
```
Comment (diracdeltas) [inline browser/ipfs/test/ipfs_service_browsertest.cc:1185]: could you also add a test that if kIPFSAutoRedirectGateway is false, it does not translate the url?

Comment (diracdeltas) [reply browser/ipfs/test/ipfs_service_browsertest.cc:1185]: same if ipfs is disabled

File: browser/net/ipfs_redirect_network_delegate_helper.cc [hunk 4]
```diff
@@ -104,25 +109,19 @@ int OnHeadersReceived_IPFSRedirectWork(
   if (ctx->ipfs_auto_fallback && !api_gateway && response_headers &&
       response_headers->GetNormalizedHeader("x-ipfs-path", &ipfs_path) &&
```
Comment (diracdeltas) [inline browser/net/ipfs_redirect_network_delegate_helper.cc:110]: does `ipfs_path` still get used anywhere?

Comment (cypt4) [reply browser/net/ipfs_redirect_network_delegate_helper.cc:110]: It is used as a marker in this case

File: browser/net/ipfs_redirect_network_delegate_helper.cc [hunk 5]
```diff
@@ -94,6 +94,11 @@ int OnHeadersReceived_IPFSRedirectWork(
     std::shared_ptr<brave::BraveRequestInfo> ctx) {
   if (!ctx->browser_context)
     return net::OK;
+
+  if (ctx->resource_type == blink::mojom::ResourceType::kSubFrame) {
```
[the remark this answers is not in the document]
Comment (cypt4) [reply browser/net/ipfs_redirect_network_delegate_helper.cc:99]: Fixed

File: components/ipfs/ipfs_utils_unittest.cc [hunk 6]
```diff
@@ -8,6 +8,7 @@
 #include <memory>
 #include <vector>
 
+#include "base/logging.h"
```
[the remark this answers is not in the document]
Comment (cypt4) [reply components/ipfs/ipfs_utils_unittest.cc]: Fixed

File: components/ipfs/ipfs_utils.cc (hunk 2 above)
Comment (cypt4) [reply components/ipfs/ipfs_utils.cc:444]: Fixed

File: browser/ipfs/test/ipfs_service_browsertest.cc (hunk 3 above)
Comment (cypt4) [reply browser/ipfs/test/ipfs_service_browsertest.cc:1301]: Fixed

File: browser/net/ipfs_redirect_network_delegate_helper.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (cypt4) [reply browser/net/ipfs_redirect_network_delegate_helper.cc:101]: Added

File: browser/net/ipfs_redirect_network_delegate_helper.cc [hunk 7]
```diff
@@ -94,6 +95,13 @@ int OnHeadersReceived_IPFSRedirectWork(
     std::shared_ptr<brave::BraveRequestInfo> ctx) {
   if (!ctx->browser_context)
     return net::OK;
+
+  // Auto-redirect gateway-like urls is enabled only for top-level frames
+  // To avoid mixed content corner cases.
```
Comment (yrliou) [inline browser/net/ipfs_redirect_network_delegate_helper.cc:100]: s/To/to

**Review by diracdeltas — APPROVED** [review]

File: components/ipfs/ipfs_utils.cc [hunk 8]
```diff
@@ -445,4 +445,47 @@ std::string GetRegistryDomainFromIPNS(const GURL& url) {
...
+  if (!url.is_valid() || !url.SchemeIsHTTPOrHTTPS()) {
+    return absl::nullopt;
+  }
+
+  std::vector<std::string> host_parts = base::SplitStringUsingSubstr(
+      url.host(), ".", base::KEEP_WHITESPACE, base::SPLIT_WANT_ALL);
+
+  if (host_parts.size() > 2 && IsValidCID(host_parts.at(0)) &&
+      host_parts.at(1) == "ipfs") {
+    GURL final_url = GURL("ipfs://" + host_parts.at(0) + url.path());
```
Comment (bridiver) [inline components/ipfs/ipfs_utils.cc:462]: you're redirecting http/https to an ipfs url? I don't think that is the correct way to handle this, you should be using url rewriting to display ipfs as the virtual url

Comment (bridiver) [reply components/ipfs/ipfs_utils.cc:462]: this is the reverse of what we do when someone enters `ipfs://` in the urlbar

Comment (cypt4) [reply components/ipfs/ipfs_utils.cc:462]: We are checking the response header at the moment

Comment (bridiver) [reply components/ipfs/ipfs_utils.cc:462]: so this redirects any gateway url? I don't think we should be doing that. If I explicitly enter an ipfs url that uses a different gateway than what I have configured, it should not redirect to a different gateway. I thought this was only to display ipfs for any urls that matched the configured gateway which can be done with the rewriter.

Comment (cypt4) [reply components/ipfs/ipfs_utils.cc:462]: > so this redirects any gateway url? I don't think we should be doing that. If I explicitly enter an ipfs url that uses a different gateway than what I have configured, it should not redirect to my configured gateway. I thought this was only to display ipfs for any urls that matched the configured gateway which can be done with the rewriter.

Nope, this is about reusing configured gateway to load IPFS resources.
> Automatically uses the configured gateway for IPFS resolutions when an IPFS gateway resource is encountered.
As for  the manual url input i'll discuss it, maybe this should be exclusion.

**Pull request merged**